[SILGen] Show a message when an unexpected enum value is switched on

jrose-apple opened the pull request:
Builds on #14724 to emit a message instead of just trapping when a switch over a non-frozen enum ends up not matching anything. If the enum is known to be an `@objc` enum, the message is

```
unexpected enum case 'MyEnum(rawValue: -42)'
```

and if it's anything else (a Swift enum, a tuple containing enums, whatever), it's a more opaque

```
unexpected enum case while switching on value of type 'MyEnum'
```

The reason for this is to avoid calling `String(describing:)` or `String(reflecting:)` an arbitrary value when the enum might conform to CustomStringConvertible and therefore *itself* have a switch that's going to fall off the end. By handling plain `@objc` enums (using a bitcast), we've at least covered the 90% case.

rdar://problem/37728359

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
For code size impact:

@swift-ci Please test compiler performance

jrose-apple commented:
And why not, maybe it'll catch a case I missed.

@swift-ci Please test source compatibility

jrose-apple commented:
Oof, forgot to update *one* test. I thought I got them all!

slavapestov commented on lib/SILGen/SILGenPattern.cpp:
```diff
@@ -2555,19 +2557,77 @@ void SILGenFunction::usingImplicitVariablesForPattern(Pattern *pattern, CaseStmt
+  assert(enumDecl->hasRawType());
+
+  // Get the enum type as an Any.Type value.
+  SILType metatypeType = SILType::getPrimitiveObjectType(
```
You’re using getLoweredType() below, using it here will allow you to pass a non-canonical metatypes without representation. A bit simpler...

jrose-apple replied:
I happen to already have canonical metatypes since they're coming from a SIL value, but I'll use getLoweredType consistently anyway.

slavapestov commented on stdlib/public/core/AssertCommon.swift:
```diff
@@ -238,3 +238,51 @@ func _undefined<T>(
+/// in release builds anyway (old apps that are run on new OSs).
+@inline(never)
+public // COMPILER_INTRINSIC
+func _diagnoseUnexpectedEnumCaseValue<RawValue>(type: Any.Type,
```
The call site will be less code if you pass in a T.Type for a new generic parameter T

jrose-apple replied:
Oh, interesting. I had assumed the opposite because there would be…more…parameters, no wait that doesn't make sense. Thanks!

slavapestov commented on stdlib/public/core/AssertCommon.swift:
```diff
@@ -238,3 +238,51 @@ func _undefined<T>(
+  let prefix: StaticString = "Fatal error"
+  let message = "unexpected enum case '\(type)(rawValue: \(rawValue))'"
+
+  prefix.withUTF8Buffer { (prefix) -> Void in
```
Does this code sequence appear in several places? Maybe extract it out

slavapestov commented on stdlib/public/core/AssertCommon.swift:
```diff
@@ -238,3 +238,51 @@ func _undefined<T>(
+/// bloats code. It doesn't take a source location because it's most important
+/// in release builds anyway (old apps that are run on new OSs).
+@inline(never)
+public // COMPILER_INTRINSIC
```
Do these have to be public? We should try using @_versioned instead

slavapestov commented on lib/SILGen/SILGenPattern.cpp:
```diff
@@ -2555,19 +2557,77 @@ void SILGenFunction::usingImplicitVariablesForPattern(Pattern *pattern, CaseStmt
+      SGF.B.createInitExistentialMetatype(loc, metatype, anyTypeType, None);
+
+  // Bitcast the enum value to its raw type. (This is only safe for @objc
+  // enums.)
```
Do we ever end up here for stringly-typed enums? Maybe assert that the raw type is trivial

jrose-apple replied:
Ah, no, we don't. Those remain structs. But I'll add the assertion.

slavapestov reviewed (approved):
LGTM with some optional suggested cleanups

jrose-apple committed:
[SILGen] Show a message when an unexpected enum value is switched on

Builds on 36eae9d4f63 to emit a message instead of just trapping
when a switch over a non-frozen enum ends up not matching anything.
If the enum is known to be an @objc enum, the message is

  unexpected enum case 'MyEnum(rawValue: -42)'

and if it's anything else (a Swift enum, a tuple containing enums,
whatever), it's a more opaque

  unexpected enum case while switching on value of type 'MyEnum'

The reason for this is to avoid calling String(describing:) or
String(reflecting:) an arbitrary value when the enum might conform to
CustomStringConvertible and therefore /itself/ have a switch that's
going to fall off the end. By handling plain @objc enums (using a
bitcast), we've at least covered the 90% case.

rdar://problem/37728359

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
(compiler performance run should keep going, and I didn't change much)

jrose-apple commented:
That is…not useful.

@swift-ci Please smoke test compiler performance

jrose-apple commented:
@swift-ci Please smoke test compiler performance

jrose-apple commented:
I'm not fully convinced by the results above and want to see what happens to code size for a larger project that uses Apple's frameworks. I'll report back soon.

graydon commented:
@swift-ci Please test compiler performance

graydon commented:
(I have attempted to fix the failure in the cperf full test, so retrying this as a "last failed run" thing)

huonw commented on lib/SILGen/SILGenPattern.cpp:
```diff
@@ -2555,19 +2557,78 @@ void SILGenFunction::usingImplicitVariablesForPattern(Pattern *pattern, CaseStmt
+  ManagedValue metatype = SGF.B.createValueMetatype(loc, metatypeType, value);
+
+  Substitution sub{switchedValueSwiftType, /*Conformances*/None};
+  auto genericArgsMap =
```
This is unused and resulting in a warning, is it unnecessary given `sub` is passed directly below?

jrose-apple replied:
Ugh, it's left over from earlier experiments. It can be removed; thanks for catching it. (`emitApplyOfLibraryIntrinsic` will do the same thing in its body, I believe.)
